### PR TITLE
Add GHA that will check links + Fix broken links

### PR DESCRIPTION
**Pull request by spier**
## :bookmark_tabs: Summary

Adds a GitHub Action (GHA) that check links in documentation files using lychee.
The markdown files that are checked so far are: `packages/mermaid/src/docs/**/*.md README.md README.zh-CN.md`

This is based on a conversation from a previous PR:
https://github.com/mermaid-js/mermaid/pull/3760#issuecomment-1304847727

## :straight_ruler: Design Decisions

- I am suggesting to run this link check once per week only, rather than for every PR. The link check can be a bit brittle, so running them for each PR can create too many false positives (I made this experience on another project).
- There were some relative links like `../config/configuration` that did not contain the `.md` extension. Therefore lychee reported these links as broken. I have added the `.md` extension for these cases
- The section about "Contributing to Documentation" has various issues, as it is still pointing to the wrong docs folder. As that is unrelated to checking broken links, I have mostly ignored these for now.

## Open Questions

- Various documentation files link to `/packages/mermaid/src/docs/config/setup/README`, which does not exist any more. Where was that content moved? => answer: auto-generated via `pnpm docs:code`
- The file `https://github.com/mermaid-js/mermaid/edit/develop/src/docs/_sidebar.md` does not exist any more. Should we link to `packages/mermaid/src/docs/.vitepress/config.ts` instead?

### :clipboard: Tasks

Make sure you

- [x] :book: have read the [contribution guidelines](https://github.com/mermaid-js/mermaid/blob/develop/CONTRIBUTING.md)
- [ ] :computer: have added unit/e2e tests (if appropriate)
- [x] :bookmark: targeted `develop` branch

### CODE REVIEW

**Comment by spier** [comment]
@aloisklink I created this based on our conversation. 

I opted to do this as a Draft PR first, so that we can work out how to best get this ready for merging.

I will leave some more comments inline to explain why i did what.

File: packages/mermaid/src/docs/community/development.md
Comment (spier) [inline packages/mermaid/src/docs/community/development.md:61]: The visible link text `mermaid-js/mermaid/src/docs` isn't correct. I didn't know what you would prefer to show here though.

File: packages/mermaid/src/docs/community/security.md
Comment (spier) [inline packages/mermaid/src/docs/community/security.md:13]: There was a type in the email address here.

Also I changed the formatting of the email addresses. 

I did not check yet if this actually works with Vitepress.
Should probably do that :)

File: packages/mermaid/src/docs/intro/index.md
Comment (spier) [inline packages/mermaid/src/docs/intro/index.md:109]: I added this to make it consistent with how all other chart types are documented.

File: packages/mermaid/src/docs/community/development.md
Comment (aloisklink) [reply packages/mermaid/src/docs/community/development.md:61]: Yep :) That one is fine! It's the main source of the documentation.

File: packages/mermaid/src/docs/community/security.md
Comment (aloisklink) [reply packages/mermaid/src/docs/community/security.md:13]: Good catch on that typo.

If it doesn't work in Vitepress, we could also make it a `mailto:` uri, see https://en.wikipedia.org/wiki/Mailto, something like:

```suggestion
You may also reach out to the team via our public Slack chat channels; however, please make sure to e-mail [security@mermaid.live](mailto:security@mermaid.live) when reporting an issue, and avoid revealing information about vulnerabilities in public as that could that could put users at risk.
```

**Review by aloisklink — COMMENTED** [review]
So far, progress looks good. Great work!

From a quick look, it seems like lychee has already caught quite a few broken links :)

---

> I am suggesting to run this link check once per week only, rather than for every PR. The link check can be a bit brittle, so running them for each PR can create too many false positives (I made this experience on another project).

My gut feeling is to leave it running on each push/PR, and if does start causing false positives, we could then disable it and just keep it running on a schedule.

But let's see what other reviewers say once this is ready to review.

---

> Various documentation files link to /packages/mermaid/src/docs/config/setup/README, which does not exist any more. Where was that content moved?

https://github.com/mermaid-js/mermaid/blob/develop/packages/mermaid/src/docs/intro/index.md
is probably the new closest thing to it.

(a processed version appears in https://github.com/mermaid-js/mermaid/blob/develop/docs/intro/index.md)

---

> The file https://github.com/mermaid-js/mermaid/edit/develop/src/docs/_sidebar.md does not exist any more. Should we link to packages/mermaid/src/docs/.vitepress/config.ts instead?

Sounds good!

---

> The section about "Contributing to Documentation" has various issues, as it is still pointing to the wrong docs folder. As that is unrelated to checking broken links, I have mostly ignored these for now.

Before we merge this, it's important that the CI passes (otherwise the link checking CI failing might hide other more important issues). It should be fine if you don't have time to update the  "Contributing to Documentation" section, but in that case, we may want to either:

a) ignore that file in CI
b) leave this PR open until a maintainer has time to fix that bit

File: packages/mermaid/src/docs/community/security.md
Comment (spier) [reply packages/mermaid/src/docs/community/security.md:13]: I have confirmed that the emails get rendered correctly and also get the `mailto:` prefix when running the docs locally. See screenshot. So we can leave this as is.

Screenshot 2022-11-07 at 10 14 35

**Comment by spier** [comment]
Figured out why the link checker could not find `/packages/mermaid/src/docs/config/setup/README`.
The content of the `setup` folder is generated from code, see `packages/mermaid/package.json`:

```
"docs:code": "typedoc src/defaultConfig.ts src/config.ts src/mermaidAPI.ts && prettier --write ./src/docs/config/setup",
```

When running that it outputs:

```
Documentation generated at ./src/docs/config/setup
src/docs/config/setup/modules/config.md 75ms
src/docs/config/setup/modules/defaultConfig.md 92ms
src/docs/config/setup/modules/mermaidAPI.md 333ms
src/docs/config/setup/README.md 4ms
```

The folder `src/docs/config/setup` is ignored by version control. See `packages/mermaid/.gitignore`.

That means that the link checker cannot check these links locally, as they don't exist as static files in the file system but rather are generated during the build.

So an interim solution for the link checker would be to ignore links below `/packages/mermaid/src/docs/config/setup`.

However there is still another mystery:

Even after generating these extra documentation files using `npm docs:code`, the documentation files I see locally are not the same as what is deployed at https://mermaid-js.github.io/mermaid/#/./Setup. I have no idea where the content that I see online comes from. Also when you try the "Edit this Page" this leads to a 404. Very strange :)

**Comment by aloisklink** [comment]
> So an interim solution for the link checker would be to ignore links below /packages/mermaid/src/docs/config/setup.

Sounds good! We could also do a `pnpm --filter mermaid run docs:build` first in GitHub Actions, but it might be simpler to leave it as something for a future PR.

> Even after generating these extra documentation files using npm docs:code, the documentation files I see locally are not the same as what is deployed at https://mermaid-js.github.io/mermaid/#/./Setup. I have no idea where the content that I see online comes from. Also when you try the "Edit this Page" this leads to a 404. Very strange :)

The old content was auto-generated from code comments by `documentationjs`, but that's unfortunately has some bugs (e.g. https://github.com/documentationjs/documentation/pull/1553), and the package is in low maintenance so they might not be fixed.

The new auto-generation code should be better in the long-term, since it supports TypeScript types, but unfortunately, a bunch of the old comments aren't compatible with the new generator, so those docs are currently pretty empty until we update those comments.

> I understood that I am not supposed to change files in the /docs folder.
> However I keep keep getting this error in the lint (18.x) GHA:
> 
> > `Changed files were transformed in src/docs but do not match the files in ../../docs. Please run 'pnpm --filter mermaid run docs:build' after making changes to src/docs to update the ../../docs directory with the transformed files.`
>
> I tried that command but it didn't resolve the issue.

I just pulled down your branch and ran `pnpm --filter mermaid run docs:build` and this is what `git diff 66c543cb8fd8bb52451ff96231976bf2949f7f22` looks like:
image

Maybe the `pnpm --filter mermaid run docs:build` script is somehow viewing the contents of your `git` repo to see whether the default `origin` is? You might have found a bug! That script was only added in https://github.com/mermaid-js/mermaid/pull/3678 (6 days ago), so thank you for being our beta tester!

As a temporary measure, you could try manually replacing all the `https://github.com/spier/mermaid` with `https://github.com/mermaid-js/mermaid` and see if that works. Otherwise, I'd be happy to just manually push to your branch once you think the PR is ready to go.

File: CONTRIBUTING.md
Comment (spier) [inline CONTRIBUTING.md:75]: This is how I got it to work at least, but would be great if somebody else could confirm this.

**Comment by spier** [comment]
@aloisklink thanks for the quick response.

I think I took this as fast as I could, as some of the remaining internals of how this all works are still unclear to me.
e.g. I don't understand how the full documentation gets published to https://mermaid-js.github.io/mermaid as it looks different from the documentation that I can generate for the project locally.

> I'd be happy to just manually push to your branch once you think the PR is ready to go.

Yes, it would be great if you could do that. Thank you!

I have marked this PR as "ready for review" 🥳

File: .github/workflows/link-checker.yml
Comment (spier) [reply .github/workflows/link-checker.yml:22]: Yes, that should work.

Actually the GITHUB_TOKEN is mostly needed to prevent being rate-limited immediately when looking up github.com links.

Feel free to commit+test your suggestions.

**Comment by sidharthv96** [comment]
> Done! I think the Mermaid team may need to investigate whether we can make pnpm run docs:verify point links to mermaid-js/mermaid even from a GitHub fork.

It should work now. I've fixed the typedoc config.

File: .github/workflows/link-checker.yml
Comment (sidharthv96) [inline .github/workflows/link-checker.yml:18]: If we have triggers on push and pull_request, why do we need a schedule?
I think we should have a push trigger on master too, as that's where the docs are actually published from.

**Review by sidharthv96 — CHANGES_REQUESTED** [review]
LGTM!

Just a small query.

File: .github/workflows/link-checker.yml
Comment (spier) [reply .github/workflows/link-checker.yml:18]: Good question @sidharthv96.

This is how I used this on another repo, with lower activity. That way even if there are no pushes/prs in a give week, the links would still get checked once.

If the mermaid repo has a lot of activity anyways, which it looks like it does, you could remove the whole `schedule` block here.

**Pull request merged**
